remote/remote.go

```diff
@@ -207,11 +207,24 @@ func newV1Image(keychain authn.Keychain, repoName string, platform imgutil.Platf
 			}
 			return nil, fmt.Errorf("connect to repo store '%s': %s", repoName, err.Error())
 		}
+		break
 	}
-
 	return image, nil
 }
 
+func isRetryable(err error) bool {
+	if err == io.EOF {
+		return true
+	}
+	if transportErr, ok := err.(*transport.Error); ok {
+		if transportErr.StatusCode == http.StatusNotFound ||
+			transportErr.StatusCode == http.StatusUnauthorized {
+			return true
+		}
+	}
```

Comment:
It's not the exact same logic since in the link you posted we also check that `len(transportErr.Errors) > 0`

Reply:
What about something like this:

```
if err != nil {
	// some code ..
	if transportErr, ok := err.(*transport.Error); ok && len(transportErr.Errors) > 0 && is40x(err) {
		return emptyImage(platform)
	}
	// more code ...
}

func isRetryable(err error) bool {
	if err == io.EOF {
		return true
	}
	return is40x(err)
}

func is40x(err error) bool {
	if transportErr, ok := err.(*transport.Error); ok {
		if transportErr.StatusCode == http.StatusNotFound ||
			transportErr.StatusCode == http.StatusUnauthorized {
			return true
		}
	}
	return false
}
```

Reply:
I don't like the idea that in the line:
```
if transportErr, ok := err.(*transport.Error); ok && len(transportErr.Errors) > 0 && is40x(err) {
```
you're running the following code twice:
```
if transportErr, ok := err.(*transport.Error); ok {
```

Maybe you can pass the `transportErr` to the `is40x` function and then it'll contain only the nested `if` in it. Something like:
```
func is40x(transportErr transport.Error) bool {
    return transportErr.StatusCode == http.StatusNotFound || transportErr.StatusCode == http.StatusUnauthorized
}
```

Reply:
When you call `transportErr, ok := err.(*transport.Error)`, this is a type assertion.
As you can read [here](https://www.sohamkamani.com/golang/type-assertions-vs-type-conversions/):
"`ok` is a boolean value which is `true` if our assertion is correct, and false otherwise."

Reply:
Marking this conversation as "unresolved", only so you'll be able to read my previous comment (otherwise, Github will hide it). Please feel free to "resolve" this conversation again.